src/session-pool.ts

```diff
@@ -820,26 +940,19 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
 
     // Only create a new session if there are more waiters than sessions already
     // being created. The current requester will be waiter number _numWaiters+1.
-    if (
-      !this.isFull &&
-      this._pending + this._pendingPrepare <= this._numWaiters
-    ) {
+    if (!this.isFull && this.totalPending <= this.totalWaiters) {
       let reads = this.options.incStep
         ? this.options.incStep
         : DEFAULTS.incStep!;
-      let writes = 0;
-      // Only create 1 write session at a time.
-      if (type === types.ReadWrite) {
-        writes++;
-        reads--;
-      }
       // Make sure we don't create more sessions than the pool should have.
-      if (reads + writes + this.size > this.options.max!) {
-        reads = this.options.max! - this.size - writes;
+      if (reads + this.size > this.options.max!) {
+        reads = this.options.max! - this.size;
       }
+      this._pending += reads;
       promises.push(
         new Promise((_, reject) => {
-          this._createSessions({reads, writes}).catch(reject);
+          this._pending -= reads;
+          this._createSessions({reads, writes: 0}).catch(reject);
```

Comment:
I can understand the confusion. Instead of determining the number of write session to create here, it is determined when the session is released into the pool. That happens on line 568 in the `release` function. That way the pool can:
1. Check whether there is anyone waiting for a session and give that session directly to that waiter, instead of first create a transaction for the session.
1. Check the actual number of write sessions in the pool at that moment, and determine whether another write session is needed or not.
1. Check whether a session was used as a write session during the last request and mark it as a session that should once again be prepared with a read/write transaction, even when the session pool configuration has the value `writes: 0`.

If `shouldBeWrite` is true and there is no process waiting for a session, the session will be marked as a write-session and a transaction will be prepared at the end of the `release` function (line 634).

Reply:
Great, thanks for the explanation @olavloite. That makes sense.